Thawing implementation in the new logic

felipeji opened the pull request:
This PR addresses Issue #13 and introduces a parallel thawing implementation to support copying objects stored in cold storage.

Cold objects are now identified and split into two copysets based on size: `smallThaw` and `largeThaw` (just as regular, hot data based in the 5 MiB threshold ). Each copyset is processed independently through its own thaw + copy map chain. Once thawed, files are copied using either Lambda (for small files) or Fargate (for large files), same that hot data.

This is implemented via two new constructs:

`thaw-small-copy-map-construct.ts`
`thaw-large-copy-map-construct.ts`

Each of these chains together `thaw-objects-map-construct` with the corresponding copy construct (`small-objects-copy-map-construct` or `large-objects-copy-map-construct`). These thawing chains are integrated into` steps-s3-copy-construct.ts,` running in parallel with hot data copiers.

In addition to the two new constructs, six other files were updated to integrate thawing and fix some issues encountered during implementation, as detailed below:

1. `dev-testsuite/test-e2e-thawing.ts` : the test timeout was increased by 2 minutes. Some thawing runs were cutting it close or failing under the original 5-minute limit.

2. `lambda/coordinate-copy-lambda.ts` new copy sets (smallThaw and largeThaw) are now created based on whether the object’s storage class is in `COLD_STORAGE_CLASSES`, and whether its size is below or above 5 MiB.
Minor TypeScript-related refactors:

- Switched to ESM-style import for path/posix
- Added non-null assertions (!) where needed to satisfy the compiler

(These two were done mostly to silence TS warnings— please, review it.)

3. `small-objects-copy-map-construct.ts` made the construct reusable by allowing `lambdaStateName` to be passed in rather than hardcoding it. There's still a bit of mixing between dynamic and hardcoded names—I might clean that up further in a follow-up ?

4&5. `thaw-objects-lambda-step-construct.ts` & `thaw-objects-map-construct.ts`: refactored to support reusable thawing logic.

Key changes:

- Introduced dynamic naming for the `lambdaFunction`

- Now using `S3CsvDistributedMap` instead of `S3JsonlDistributedMap`, since we're now reading thaw instructions directly from JSONL (via the `copySets.smallThaw` / `largeThaw` input)

6. steps-s3-copy-construct.ts: Integrated calls to the new thawing constructs using the same naming conventions and parallel execution logic as for hot data copiers.
NOTE: I’m manually granting invoke permissions to the thawing Lambdas (line ~350). It works, but I’m not entirely sure it’s the cleanest/correct way to do it.

andrewpatto commented on packages/steps-s3-copy/src/lib/thaw-large-copy-map-construct.ts:
```diff
@@ -0,0 +1,71 @@
+      inputPath: props.inputPath,
+      mapStateName: id,
+    });
+
```
What would be the ramifications if one thaw completed quickly - and the others took 48 hours?

Might the quick thaw "re-expire"??

andrewpatto replied:
I thought potentially for (large or small) thawing we would have only one distributed map - but that _within_ that map would be a single chain for each object doing the thaw->wait->copy loop

felipeji commented:
The previous implementation of thawing was not optimal, as it performed all thawing operations before initiating the copy tasks instead of in the same map, which is  inefficiency.

To address this, I’ve moved the thawing logic into the copy map constructs themselves—SmallObjectsCopyMapConstruct and CopyMapConstruct, so that thawing is now triggered via the `thaw` flag when they are instantiated. This new approach allows each object to be thawed just before it is copied. 

The logic is now implemented in:

`packages/steps-s3-copy/src/lib/copy-map-construct.ts`
`packages/steps-s3-copy/src/lib/small-copy-map-construct.ts`

So, as a result, the following thawing-specific constructs are no longer needed and have been removed:

```
packages/steps-s3-copy/src/lib/thaw-large-copy-map-construct.ts
packages/steps-s3-copy/src/lib/thaw-objects-map-construct.ts
packages/steps-s3-copy/src/lib/thaw-small-copy-map-construct.ts
```

Other changes include:

In `packages/steps-s3-copy/lambda/can-read-objects-lambda/can-read-objects-lambda.ts,` the original event is now passed downstream so that it remains available to the copy step after thawing.

In `packages/steps-s3-copy/src/lib/thaw-lambda-step-construct.ts`, I added outputPath: "$.Payload" to handling of InputPath in downstream copy steps.

I also cleaned up the state and step names for clarity, while still using CamelCase for naming. We could consider setting more human-readable names for those steps in the future (?)

andrewpatto commented on packages/steps-s3-copy/lambda/can-read-objects-lambda/can-read-objects-lambda.ts:
```diff
@@ -145,10 +145,11 @@ export async function handler(event: ThawObjectsEvent) {
-  return {};
+  }
+  // Return original input so downstream steps (e.g. ECS copy) get expected fields
+  return event;
```
Just as a note - you can also pass through to downstream steps with config in the Steps. As in - you can ignore the the return value of the lambda and have Steps itself pass through values.

No need to change here - is really possibly to do with us progressing to full JSONata at some point

felipeji replied:
I need to learn more on this. I was dealing quit a bit with this, and was the best solution I found (or at least the one I actually understood) cos the copy step brokes when trying to pass the inputs needed for thawing. I’m happy to work on trying the full implementation JSONata, so I can learn.

andrewpatto commented on packages/steps-s3-copy/src/lib/copy-map-construct.ts:
```diff
@@ -71,6 +75,41 @@ export class CopyMapConstruct extends Construct {
+    } else {
+      entryState = delayStep;
+    }
+
```
nice!

andrewpatto commented on packages/steps-s3-copy/src/lib/copy-map-construct.ts:
```diff
@@ -42,6 +43,9 @@ type Props = {
   readonly taskDefinition: TaskDefinition;
   readonly containerDefinition: ContainerDefinition;
+
+  readonly thaw?: boolean;
```
I would clarfiy the name of this - to make it clear that this `thaw` changes the Construct and is not dynamic behaviour.

So change it to be a verb _about_ the construct. Maybe `addThawStep`?

felipeji replied:
Done!

andrewpatto commented on packages/steps-s3-copy/src/lib/copy-map-construct.ts:
```diff
@@ -147,17 +186,36 @@ export class CopyMapConstruct extends Construct {
+        intelligentTieringDeepArchiveThawSpeed: props.aggressiveTimes
+          ? "Standard"
+          : "Bulk",
+      },
```
I will put in a new issue - so not for this PR - but we will eventually want these parameters to be allowed to be specified by the invoker of the copy (so possibly bringing this entire data structure up to the Steps invoke parameter level)
